File: packages/nhost-js/src/nhost.ts
import { generateServiceUrl } from './';
import {
  type Client as AuthClient,
  createAPIClient as createAuthClient,
} from './auth';
import {
  type AdminSessionOptions,
  attachAccessTokenMiddleware,
  type ChainFunction,
  sessionRefreshMiddleware,
  updateSessionFromResponseMiddleware,
  withAdminSessionMiddleware,
} from './fetch';
import {
  createAPIClient as createFunctionsClient,
  type Client as FunctionsClient,
} from './functions';
import {
  createAPIClient as createGraphQLClient,
  type Client as GraphQLClient,
} from './graphql';
import {
  detectStorage,
  refreshSession,
  type Session,
  SessionStorage,
  type SessionStorageBackend,
} from './session/';
import {
  createAPIClient as createStorageClient,
  type Client as StorageClient,
} from './storage';

/**
 * Configuration function that receives all clients and can configure them
 * (e.g., by attaching middleware, setting up interceptors, etc.)
 */
export type ClientConfigurationFn = (clients: {
  auth: AuthClient;
  storage: StorageClient;
  graphql: GraphQLClient;
  functions: FunctionsClient;
  sessionStorage: SessionStorage;
}) => void;

/**
 * Built-in configuration for client-side applications.
 * Includes automatic session refresh, token attachment, and session updates.
 */
export const withClientSideSessionMiddleware: ClientConfigurationFn = ({
  auth,
  storage,
  graphql,
  functions,
  sessionStorage,
}) => {
  const mwChain: ChainFunction[] = [
    sessionRefreshMiddleware(auth, sessionStorage),
    updateSessionFromResponseMiddleware(sessionStorage),
    attachAccessTokenMiddleware(sessionStorage),
  ];

  for (const mw of mwChain) {
    auth.pushChainFunction(mw);
    storage.pushChainFunction(mw);
    graphql.pushChainFunction(mw);
    functions.pushChainFunction(mw);
  }
};

/**
 * Built-in configuration for server-side applications.
 * Includes token attachment and session updates, but NOT automatic session refresh
 * to prevent race conditions in server contexts.
 */
export const withServerSideSessionMiddleware: ClientConfigurationFn = ({
  auth,
  storage,
  graphql,
  functions,
  sessionStorage,
}) => {
  const mwChain: ChainFunction[] = [
    updateSessionFromResponseMiddleware(sessionStorage),
    attachAccessTokenMiddleware(sessionStorage),
  ];

  for (const mw of mwChain) {
    auth.pushChainFunction(mw);
    storage.pushChainFunction(mw);
    graphql.pushChainFunction(mw);
    functions.pushChainFunction(mw);
  }
};

/**
 * Configuration for admin clients with elevated privileges.
 * Applies admin session middleware to storage, graphql, and functions clients only.
 *
 * **Security Warning**: Never use this in client-side code. Admin secrets grant
 * unrestricted access to your entire database.
 *
 * @param adminSession - Admin session options including admin secret, role, and session variables
 * @returns Configuration function that sets up admin middleware
 */
export function withAdminSession(
  adminSession: AdminSessionOptions,
): ClientConfigurationFn {
  return ({ storage, graphql, functions }) => {
    const adminMiddleware = withAdminSessionMiddleware(adminSession);

    storage.pushChainFunction(adminMiddleware);
    graphql.pushChainFunction(adminMiddleware);
    functions.pushChainFunction(adminMiddleware);
  };
}

/**
 * Configuration for adding custom chain functions to all clients.
 * Useful for adding custom middleware like logging, caching, or custom headers.
 *
 * @param chainFunctions - Array of chain functions to apply to all clients
 * @returns Configuration function that sets up custom middleware
 */
export function withChainFunctions(
  chainFunctions: ChainFunction[],
): ClientConfigurationFn {
  return ({ auth, storage, graphql, functions }) => {
    for (const mw of chainFunctions) {
      auth.pushChainFunction(mw);
      storage.pushChainFunction(mw);
      graphql.pushChainFunction(mw);
      functions.pushChainFunction(mw);
    }
  };
}

/**
 * Main client class that provides unified access to all Nhost services.
 * This class serves as the central interface for interacting with Nhost's
 * authentication, storage, GraphQL, and serverless functions capabilities.
 */
export class NhostClient {
  /**
   * Authentication client providing methods for user sign-in, sign-up, and session management.
   * Use this client to handle all authentication-related operations.
   */
  auth: AuthClient;

  /**
   * Storage client providing methods for file operations (upload, download, delete).
   * Use this client to manage files in your Nhost storage.
   */
  storage: StorageClient;

  /**
   * GraphQL client providing methods for executing GraphQL operations against your Hasura backend.
   * Use this client to query and mutate data in your database through GraphQL.
   */
  graphql: GraphQLClient;

  /**
   * Functions client providing methods for invoking serverless functions.
   * Use this client to call your custom serverless functions deployed to Nhost.
   */
  functions: FunctionsClient;

  /**
   * Storage implementation used for persisting session information.
   * This handles saving, retrieving, and managing authentication sessions across requests.
   */
  sessionStorage: SessionStorage;

  /**
   * Create a new Nhost client. This constructor is reserved for advanced use cases.
   * For typical usage, use [createClient](#createclient) or [createServerClient](#createserverclient) instead.
   *
   * @param auth - Authentication client instance
   * @param storage - Storage client instance
   * @param graphql - GraphQL client instance
   * @param functions - Functions client instance
   * @param sessionStorage - Storage implementation for session persistence
   */
  constructor(
    auth: AuthClient,
    storage: StorageClient,
    graphql: GraphQLClient,
    functions: FunctionsClient,
    sessionStorage: SessionStorage,
  ) {
    this.auth = auth;
    this.storage = storage;
    this.graphql = graphql;
    this.functions = functions;
    this.sessionStorage = sessionStorage;
  }

  /**
   * Get the current session from storage.
   * This method retrieves the authenticated user's session information if one exists.
   *
   * @returns The current session or null if no session exists
   *
   * @example
   * ```ts
   * const session = nhost.getUserSession();
   * if (session) {
   *   console.log('User is authenticated:', session.user.id);
   * } else {
   *   console.log('No active session');
   * }
   * ```
   */
  getUserSession(): Session | null {
    return this.sessionStorage.get();
  }

  /**
   * Refresh the session using the current refresh token
   * in the storage and update the storage with the new session.
   *
   * This method can be used to proactively refresh tokens before they expire
   * or to force a refresh when needed.
   *
   * @param marginSeconds - The number of seconds before the token expiration to refresh the session. If the token is still valid for this duration, it will not be refreshed. Set to 0 to force the refresh.
   *
   * @returns The new session or null if there is currently no session or if refresh fails
   *
   * @example
   * ```ts
   * // Refresh token if it's about to expire in the next 5 minutes
   * const refreshedSession = await nhost.refreshSession(300);
   *
   * // Force refresh regardless of current token expiration
   * const forcedRefresh = await nhost.refreshSession(0);
   * ```
   */
  async refreshSession(marginSeconds = 60): Promise<Session | null> {
    return refreshSession(this.auth, this.sessionStorage, marginSeconds);
  }

  /**
   * Clear the session from storage.
   *
   * This method removes the current authentication session, effectively logging out the user.
   * Note that this is a client-side operation and doesn't invalidate the refresh token on
   * the server, which can be done with `nhost.auth.signOut({refreshToken: session.refreshTokenId})`.
   * If the middle `updateSessionFromResponseMiddleware` is used, the session will be removed
   * from the storage automatically and calling this method is not necessary.
   *
   * @example
   * ```ts
   * // Log out the user
   * nhost.clearSession();
   * ```
   */
  clearSession(): void {
    this.sessionStorage.remove();
  }
}

/**
 * Configuration options for creating an Nhost client
 */
export interface NhostClientOptions {
  /**
   * Nhost project subdomain (e.g., 'abcdefgh'). Used to construct the base URL for services for the Nhost cloud.
   */
  subdomain?: string;

  /**
   * Nhost region (e.g., 'eu-central-1'). Used to construct the base URL for services for the Nhost cloud.
   */
  region?: string;

  /**
   * Complete base URL for the auth service (overrides subdomain/region)
   */
  authUrl?: string;

  /**
   * Complete base URL for the storage service (overrides subdomain/region)
   */
  storageUrl?: string;

  /**
   * Complete base URL for the GraphQL service (overrides subdomain/region)
   */
  graphqlUrl?: string;

  /**
   * Complete base URL for the functions service (overrides subdomain/region)
   */
  functionsUrl?: string;

  /**
   * Storage backend to use for session persistence. If not provided, the SDK will
   * default to localStorage in the browser or memory in other environments.
   */
  storage?: SessionStorageBackend;

  /**
   * Configuration functions to be applied to the client after initialization.
   * These functions receive all clients and can attach middleware or perform other setup.
   */
  configure?: ClientConfigurationFn[];
}

/**
 * Creates and configures a new Nhost client instance with custom configuration.
 *
 * This is the main factory function for creating Nhost clients. It instantiates
 * all service clients (auth, storage, graphql, functions) and applies the provided
 * configuration functions to set up middleware and other customizations.
 *
 * @param options - Configuration options for the client
 * @returns A configured Nhost client
 *
 * @example
 * ```ts
 * // Create a basic client with no middleware
 * const nhost = createNhostClient({
 *   subdomain: 'abcdefgh',
 *   region: 'eu-central-1',
 *   configure: []
 * });
 *
 * // Create a client with custom configuration
 * const nhost = createNhostClient({
 *   subdomain: 'abcdefgh',
 *   region: 'eu-central-1',
 *   configure: [
 *     withClientSideSessionMiddleware,
 *     withChainFunctions([customLoggingMiddleware])
 *   ]
 * });
 *
 * // Create an admin client
 * const nhost = createNhostClient({
 *   subdomain,
 *   region,
 *   configure: [
 *     withAdminSession({
 *       adminSecret: "nhost-admin-secret",
 *       role: "user",
 *       sessionVariables: {
 *         "user-id": "54058C42-51F7-4B37-8B69-C89A841D2221",
 *       },
 *     }),
 *   ],
 * });

 * ```
 */
export function createNhostClient(
  options: NhostClientOptions = {},
): NhostClient {
  const {
    subdomain,
    region,
    authUrl,
    storageUrl,
    graphqlUrl,
    functionsUrl,
    storage = detectStorage(),
    configure = [],
  } = options;

  const sessionStorage = new SessionStorage(storage);

  // Determine base URLs for each service
  const authBaseUrl = generateServiceUrl('auth', subdomain, region, authUrl);
  const storageBaseUrl = generateServiceUrl(
    'storage',
    subdomain,
    region,
    storageUrl,
  );
  const graphqlBaseUrl = generateServiceUrl(
    'graphql',
    subdomain,
    region,
    graphqlUrl,
  );
  const functionsBaseUrl = generateServiceUrl(
    'functions',
    subdomain,
    region,
    functionsUrl,
  );

  // Create all clients
  const auth = createAuthClient(authBaseUrl);
  const storageClient = createStorageClient(storageBaseUrl, []);
  const graphqlClient = createGraphQLClient(graphqlBaseUrl, []);
  const functionsClient = createFunctionsClient(functionsBaseUrl, []);

  // Apply configuration functions
  for (const configFn of configure) {
    configFn({
      auth,
      storage: storageClient,
      graphql: graphqlClient,
      functions: functionsClient,
      sessionStorage,
    });
  }

  // Return an initialized NhostClient
  return new NhostClient(
    auth,
    storageClient,
    graphqlClient,
    functionsClient,
    sessionStorage,
  );
}

/**
 * Creates and configures a new Nhost client instance optimized for client-side usage.
 *
 * This helper method instantiates a fully configured Nhost client by:
 * - Instantiating the various service clients (auth, storage, functions and graphql)
 * - Auto-detecting and configuring an appropriate session storage (localStorage in browsers, memory otherwise)
 * - Setting up a sophisticated middleware chain for seamless authentication management:
 *   - Automatically refreshing tokens before they expire
 *   - Attaching authorization tokens to all service requests
 *   - Updating the session storage when new tokens are received
 *
 * This method includes automatic session refresh middleware, making it ideal for
 * client-side applications where long-lived sessions are expected.
 *
 * @param options - Configuration options for the client
 * @returns A configured Nhost client
 *
 * @example
 * ```ts
 * // Create client using Nhost cloud default URLs
 * const nhost = createClient({
 *   subdomain: 'abcdefgh',
 *   region: 'eu-central-1'
 * });
 *
 * // Create client with custom service URLs
 * const customNhost = createClient({
 *   authUrl: 'https://auth.example.com',
 *   storageUrl: 'https://storage.example.com',
 *   graphqlUrl: 'https://graphql.example.com',
 *   functionsUrl: 'https://functions.example.com'
 * });
 *
 * // Create client using cookies for storing the session
 * import { CookieStorage } from "@nhost/nhost-js/session";
 *
 * const nhost = createClient({
 *   subdomain: 'abcdefgh',
 *   region: 'eu-central-1',
 *   storage: new CookieStorage({
 *      secure: import.meta.env.ENVIRONMENT === 'production',
 *   })
 * });
 *
 * // Create client with additional custom middleware
 * const nhost = createClient({
 *   subdomain: 'abcdefgh',
 *   region: 'eu-central-1',
 *   configure: [customLoggingMiddleware]
 * });
 * ```
 */
export function createClient(options: NhostClientOptions = {}): NhostClient {
  const storage = options.storage ?? detectStorage();

  return createNhostClient({
    ...options,
    storage,
    configure: [withClientSideSessionMiddleware, ...(options.configure ?? [])],
  });
}

export interface NhostServerClientOptions extends NhostClientOptions {
  /**
   * Storage backend to use for session persistence in server environments.
   * Unlike the base options, this field is required for server-side usage
   * as the SDK cannot auto-detect an appropriate storage mechanism.
   */
  storage: SessionStorageBackend;
}

/**
 * Creates and configures a new Nhost client instance optimized for server-side usage.
 *
 * This helper method instantiates a fully configured Nhost client specifically designed for:
 * - Server components (in frameworks like Next.js or Remix)
 * - API routes and middleware
 * - Backend services and server-side rendering contexts
 *
 * Key differences from the standard client:
 * - Requires explicit storage implementation (must be provided)
 * - Disables automatic session refresh middleware (to prevent race conditions in server contexts)
 * - Still attaches authorization tokens and updates session storage from responses
 *
 * The server client is ideal for short-lived request contexts where session tokens
 * are passed in (like cookie-based authentication flows) and automatic refresh
 * mechanisms could cause issues with concurrent requests.
 *
 * @param options - Configuration options for the server client (requires storage implementation)
 * @returns A configured Nhost client optimized for server-side usage
 *
 * @example
 * ```ts
 * // Example with cookie storage for Next.js API route or server component
 * import { cookies } from 'next/headers';
 *
 * const nhost = createServerClient({
 *   region: process.env["NHOST_REGION"] || "local",
 *   subdomain: process.env["NHOST_SUBDOMAIN"] || "local",
 *   storage: {
 *     // storage compatible with Next.js server components
 *     get: (): Session | null => {
 *       const s = cookieStore.get(key)?.value || null;
 *       if (!s) {
 *         return null;
 *       }
 *       const session = JSON.parse(s) as Session;
 *       return session;
 *     },
 *     set: (value: Session) => {
 *       cookieStore.set(key, JSON.stringify(value));
 *     },
 *     remove: () => {
 *       cookieStore.delete(key);
 *     },
 *   },
 * });
 *
 * // Example with cookie storage for Next.js middleware
 * const nhost = createServerClient({
 *   region: process.env["NHOST_REGION"] || "local",
 *   subdomain: process.env["NHOST_SUBDOMAIN"] || "local",
 *   storage: {
 *     // storage compatible with Next.js middleware
 *     get: (): Session | null => {
 *       const raw = request.cookies.get(key)?.value || null;
 *       if (!raw) {
 *         return null;
 *       }
 *       const session = JSON.parse(raw) as Session;
 *       return session;
 *     },
 *     set: (value: Session) => {
 *       response.cookies.set({
 *         name: key,
 *         value: JSON.stringify(value),
 *         path: "/",
 *         httpOnly: false, //if set to true we can't access it in the client
 *         secure: process.env.NODE_ENV === "production",
 *         sameSite: "lax",
 *         maxAge: 60 * 60 * 24 * 30, // 30 days in seconds
 *       });
 *     },
 *     remove: () => {
 *       response.cookies.delete(key);
 *     },
 *   },
 * });
 *
 * // Example for express reading session from a cookie
 *
 * import express, { Request, Response } from "express";
 * import cookieParser from "cookie-parser";
 *
 * app.use(cookieParser());
 *
 * const nhostClientFromCookies = (req: Request) => {
 *   return createServerClient({
 *     subdomain: "local",
 *     region: "local",
 *     storage: {
 *       get: (): Session | null => {
 *         const s = req.cookies.nhostSession || null;
 *         if (!s) {
 *           return null;
 *         }
 *         const session = JSON.parse(s) as Session;
 *         return session;
 *       },
 *       set: (_value: Session) => {
 *         throw new Error("It is easier to handle the session in the client");
 *       },
 *       remove: () => {
 *         throw new Error("It is easier to handle the session in the client");
 *       },
 *     },
 *   });
 * };
 *
 * // Example with additional custom middleware
 * const nhost = createServerClient({
 *   region: process.env["NHOST_REGION"] || "local",
 *   subdomain: process.env["NHOST_SUBDOMAIN"] || "local",
 *   storage: myStorage,
 *   configure: [customLoggingMiddleware]
 * });
 * ```
 */
export function createServerClient(
  options: NhostServerClientOptions,
): NhostClient {
  return createNhostClient({
    ...options,
    configure: [withServerSideSessionMiddleware, ...(options.configure ?? [])],
  });
}

File: packages/nhost-python/src/nhost/nhost.py
"""Top-level Nhost client and factory functions."""

from __future__ import annotations

import re
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import Any, Literal, cast

import httpx

from . import auth as auth_module
from . import functions as functions_module
from . import graphql as graphql_module
from . import storage as storage_module
from .fetch import (
    AdminSessionOptions,
    Middleware,
    attach_access_token_middleware,
    session_refresh_middleware,
    update_session_from_response_middleware,
    with_admin_session_middleware,
)
from .session import MemoryStorage, SessionStorage, SessionStorageBackend, StoredSession
from .session.refresh import refresh_session

ServiceType = Literal["auth", "storage", "graphql", "functions"]
ClientConfiguration = Callable[["ConfigureContext"], None]

_DEFAULT_REFRESH_MARGIN_SECONDS = 60
_DEFAULT_HTTP_TIMEOUT = httpx.Timeout(connect=10.0, read=300.0, write=300.0, pool=60.0)
_CLOUD_HOST_LABEL = re.compile(r"[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?")


def _validate_cloud_host_label(name: str, value: str) -> None:
    if _CLOUD_HOST_LABEL.fullmatch(value) is None:
        raise ValueError(
            f"{name} must be a valid DNS label: 1-63 ASCII letters, digits, or hyphens, "
            "starting and ending with a letter or digit"
        )


def generate_service_url(
    service_type: ServiceType,
    *,
    subdomain: str | None = None,
    region: str | None = None,
    custom_url: str | None = None,
) -> str:
    """Build a normalized service URL.

    An explicit ``custom_url`` takes precedence. Otherwise ``subdomain`` and
    ``region`` must be supplied together; omitting both selects the local Nhost
    development URL.

    >>> generate_service_url("auth", subdomain="demo", region="eu-central-1")
    'https://demo.auth.eu-central-1.nhost.run/v1'
    >>> generate_service_url("graphql")
    'https://local.graphql.local.nhost.run/v1'
    >>> generate_service_url("functions", custom_url="http://localhost:1337/v1/")
    'http://localhost:1337/v1'
    """
    if custom_url is not None:
        try:
            parsed = httpx.URL(custom_url)
        except httpx.InvalidURL as error:
            raise ValueError(f"invalid custom URL for {service_type}: {error}") from error
        if parsed.scheme not in {"http", "https"} or not parsed.host or parsed.userinfo:
            raise ValueError(
                f"custom URL for {service_type} must supply an explicit HTTP or HTTPS scheme, "
                "include a host, and omit user information"
            )
        return str(parsed).rstrip("/")

    if subdomain is not None:
        _validate_cloud_host_label("subdomain", subdomain)
    if region is not None:
        _validate_cloud_host_label("region", region)
    if (subdomain is None) != (region is None):
        raise ValueError("subdomain and region must be supplied together")
    if subdomain is None:
        return f"https://local.{service_type}.local.nhost.run/v1"
    return f"https://{subdomain}.{service_type}.{region}.nhost.run/v1"


@dataclass(slots=True)
class ConfigureContext:
    """Clients and session storage passed to a configuration callback."""

    auth: auth_module.AuthClient
    refresh_auth: auth_module.AuthClient
    storage: storage_module.StorageClient
    graphql: graphql_module.Client
    functions: functions_module.Client
    session_storage: SessionStorage


def _attach_access_token_to_each_service(ctx: ConfigureContext) -> None:
    ctx.auth.add_middleware(attach_access_token_middleware(ctx.session_storage, ctx.auth.base_url))
    ctx.storage.add_middleware(
        attach_access_token_middleware(ctx.session_storage, ctx.storage.base_url)
    )
    ctx.graphql.add_middleware(
        attach_access_token_middleware(ctx.session_storage, ctx.graphql.base_url)
    )
    ctx.functions.add_middleware(
        attach_access_token_middleware(ctx.session_storage, ctx.functions.base_url)
    )


def with_client_side_session_middleware(ctx: ConfigureContext) -> None:
    """Enable automatic refresh, token attachment, and session capture."""
    _apply(
        ctx,
        [
            session_refresh_middleware(ctx.refresh_auth, ctx.session_storage),
            update_session_from_response_middleware(ctx.session_storage, ctx.auth.base_url),
        ],
    )
    _attach_access_token_to_each_service(ctx)


def with_server_side_session_middleware(ctx: ConfigureContext) -> None:
    """Enable token attachment and session capture without automatic refresh."""
    _apply(
        ctx,
        [update_session_from_response_middleware(ctx.session_storage, ctx.auth.base_url)],
    )
    _attach_access_token_to_each_service(ctx)


def with_admin_session(options: AdminSessionOptions) -> ClientConfiguration:
    """Apply admin credentials to Storage, GraphQL, and Functions requests.

    Never use an admin secret in client-side code.
    """

    def configure(ctx: ConfigureContext) -> None:
        ctx.storage.add_middleware(with_admin_session_middleware(options, ctx.storage.base_url))
        ctx.graphql.add_middleware(with_admin_session_middleware(options, ctx.graphql.base_url))
        ctx.functions.add_middleware(with_admin_session_middleware(options, ctx.functions.base_url))

    return configure


def with_middleware(middleware: Sequence[Middleware]) -> ClientConfiguration:
    """Apply custom HTTP middleware to every service client."""
    chain = list(middleware)

    def configure(ctx: ConfigureContext) -> None:
        _apply(ctx, chain)

    return configure


def _apply(ctx: ConfigureContext, chain: Sequence[Middleware]) -> None:
    for middleware in chain:
        ctx.auth.add_middleware(middleware)
        ctx.storage.add_middleware(middleware)
        ctx.graphql.add_middleware(middleware)
        ctx.functions.add_middleware(middleware)


class _PerRequestTimeoutClient:
    """Apply an SDK timeout while delegating transport ownership to another client."""

    def __init__(
        self,
        client: httpx.AsyncClient,
        timeout: httpx.Timeout | float | None,
    ) -> None:
        self._client = client
        self._timeout = timeout

    def build_request(self, *args: Any, **kwargs: Any) -> httpx.Request:
        kwargs.setdefault("timeout", self._timeout)
        return self._client.build_request(*args, **kwargs)

    async def send(self, request: httpx.Request, **kwargs: Any) -> httpx.Response:
        return await self._client.send(request, **kwargs)


class NhostClient:
    """Unified asynchronous access to Nhost services and session state."""

    def __init__(
        self,
        auth: auth_module.AuthClient,
        refresh_auth: auth_module.AuthClient,
        storage: storage_module.StorageClient,
        graphql: graphql_module.Client,
        functions: functions_module.Client,
        session_storage: SessionStorage,
        http_client: httpx.AsyncClient,
        *,
        owns_http: bool = True,
    ) -> None:
        self.auth = auth
        self._refresh_auth = refresh_auth
        self.storage = storage
        self.graphql = graphql
        self.functions = functions
        self.session_storage = session_storage
        self._http = http_client
        self._owns_http = owns_http

    async def get_session(self) -> StoredSession | None:
        """Return the current session, if one is stored."""
        return await self.session_storage.get()

    async def refresh_session(
        self, margin_seconds: int = _DEFAULT_REFRESH_MARGIN_SECONDS
    ) -> StoredSession | None:
        """Refresh the session when it is close to expiry."""
        return await refresh_session(self._refresh_auth, self.session_storage, margin_seconds)

    async def clear_session(self) -> None:
        """Remove the current session without making a sign-out request."""
        await self.session_storage.remove()

    async def aclose(self) -> None:
        """Close the internally owned HTTP connection pool, if any."""
        if self._owns_http:
            await self._http.aclose()

    async def __aenter__(self) -> NhostClient:
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.aclose()


def create_nhost_client(  # noqa: PLR0913 - explicit keyword API is intentional
    *,
    subdomain: str | None = None,
    region: str | None = None,
    auth_url: str | None = None,
    storage_url: str | None = None,
    graphql_url: str | None = None,
    functions_url: str | None = None,
    session_storage: SessionStorageBackend | None = None,
    http_client: httpx.AsyncClient | None = None,
    configure: Sequence[ClientConfiguration] = (),
    timeout: httpx.Timeout | float | None = _DEFAULT_HTTP_TIMEOUT,
) -> NhostClient:
    """Create a bare Nhost client from explicit keyword configuration.

    ``timeout`` applies to each SDK-built request, including when ``http_client``
    is supplied. A timeout set explicitly for one request takes precedence.
    """
    if (subdomain is None) != (region is None):
        raise ValueError("subdomain and region must be supplied together")

    backend = session_storage if session_storage is not None else MemoryStorage()
    sessions = SessionStorage(backend)
    http = http_client if http_client is not None else httpx.AsyncClient(timeout=timeout)
    request_http = (
        http
        if http_client is None
        else cast(httpx.AsyncClient, _PerRequestTimeoutClient(http, timeout))
    )
    resolved_auth_url = generate_service_url(
        "auth", subdomain=subdomain, region=region, custom_url=auth_url
    )

    auth = auth_module.AuthClient(resolved_auth_url, http_client=request_http)
    refresh_auth = auth_module.AuthClient(resolved_auth_url, http_client=request_http)
    storage = storage_module.StorageClient(
        generate_service_url("storage", subdomain=subdomain, region=region, custom_url=storage_url),
        http_client=request_http,
    )
    graphql = graphql_module.Client(
        generate_service_url("graphql", subdomain=subdomain, region=region, custom_url=graphql_url),
        http_client=request_http,
    )
    functions = functions_module.Client(
        generate_service_url(
            "functions", subdomain=subdomain, region=region, custom_url=functions_url
        ),
        http_client=request_http,
    )

    ctx = ConfigureContext(auth, refresh_auth, storage, graphql, functions, sessions)
    for configure_client in configure:
        configure_client(ctx)

    return NhostClient(
        auth,
        refresh_auth,
        storage,
        graphql,
        functions,
        sessions,
        http,
        owns_http=http_client is None,
    )


def create_client(  # noqa: PLR0913 - explicit keyword API is intentional
    *,
    subdomain: str | None = None,
    region: str | None = None,
    auth_url: str | None = None,
    storage_url: str | None = None,
    graphql_url: str | None = None,
    functions_url: str | None = None,
    session_storage: SessionStorageBackend | None = None,
    http_client: httpx.AsyncClient | None = None,
    configure: Sequence[ClientConfiguration] = (),
    timeout: httpx.Timeout | float | None = _DEFAULT_HTTP_TIMEOUT,
) -> NhostClient:
    """Create an application client with automatic session management.

    ``timeout`` applies to SDK requests even when ``http_client`` is supplied.

    >>> import asyncio, uuid
    >>> from nhost.auth import SignUpEmailPasswordRequest
    >>> async def main() -> str | None:
    ...     async with create_client(subdomain="local", region="local") as nhost:
    ...         await nhost.auth.sign_up_email_password(
    ...             body=SignUpEmailPasswordRequest(
    ...                 email=f"ada-{uuid.uuid4()}@example.com",
    ...                 password=str(uuid.uuid4()),
    ...             )
    ...         )
    ...         session = await nhost.get_session()
    ...         return (session.decoded_token.hasura_claims or {}).get(
    ...             "x-hasura-default-role"
    ...         )
    >>> asyncio.run(main())
    'user'
    """
    return create_nhost_client(
        subdomain=subdomain,
        region=region,
        auth_url=auth_url,
        storage_url=storage_url,
        graphql_url=graphql_url,
        functions_url=functions_url,
        session_storage=session_storage,
        http_client=http_client,
        configure=(*configure, with_client_side_session_middleware),
        timeout=timeout,
    )


def create_server_client(  # noqa: PLR0913 - explicit keyword API is intentional
    *,
    session_storage: SessionStorageBackend,
    subdomain: str | None = None,
    region: str | None = None,
    auth_url: str | None = None,
    storage_url: str | None = None,
    graphql_url: str | None = None,
    functions_url: str | None = None,
    http_client: httpx.AsyncClient | None = None,
    configure: Sequence[ClientConfiguration] = (),
    timeout: httpx.Timeout | float | None = _DEFAULT_HTTP_TIMEOUT,
) -> NhostClient:
    """Create a server client with explicit per-user session storage.

    ``timeout`` applies to SDK requests even when ``http_client`` is supplied.
    """
    return create_nhost_client(
        subdomain=subdomain,
        region=region,
        auth_url=auth_url,
        storage_url=storage_url,
        graphql_url=graphql_url,
        functions_url=functions_url,
        session_storage=session_storage,
        http_client=http_client,
        configure=(*configure, with_server_side_session_middleware),
        timeout=timeout,
    )
